scripts/data_validation.py

```diff
@@ -0,0 +1,674 @@
+    generation_pypsa.name = "PyPSA data"
+    generation_pypsa = generation_pypsa.to_frame()
+    return generation_pypsa
+
+
+def get_generation_capacity_ember_detail(data, three_country_code, year):
+    """
+    Get electricity generation by fuel type for a given country and year from Ember data.
+
+    Args:
+        data (pd.DataFrame): Ember data.
+        three_country_code (str): Country code (ISO 3166-1 alpha-3).
+        year (int): Year of interest.
+
+    Returns:
+        pd.DataFrame: Electricity generation by fuel type.
+    """
+    generation_ember = data[
+        (data["Category"] == "Electricity generation")
+        & (data["Country code"] == three_country_code)
+        & (data["Year"] == year)
+        & (data["Subcategory"] == "Fuel")
+        & (data["Unit"] == "TWh")
+    ][["Variable", "Value"]].reset_index(drop=True)
+
+    # Drop irrelevant rows
+    drop_row = ["Other Renewables"]
+    generation_ember = generation_ember[~generation_ember["Variable"].isin(
+        drop_row)]
+
+    # Standardize fuel types
+    generation_ember = generation_ember.replace({
+        "Gas": "Natural gas",
+        "Bioenergy": "Biomass",
+        # "Coal": "Fossil fuels",
+        # "Other Fossil": "Fossil fuels"
+    })
+
+    # Group by fuel type
+    generation_ember = generation_ember.groupby("Variable").sum()
+    generation_ember.loc["Load shedding"] = 0.0
+    generation_ember.columns = ["Ember data"]
+
+    return generation_ember
+
+
+def plot_demand_validation(demand_ember, pypsa_demand, EIA_demand, horizon, country_code):
+    plt.figure(figsize=(8, 6))  # Set figure size
+    bars = plt.bar(["Ember", "PyPSA", "EIA"], [
+                   demand_ember, pypsa_demand, EIA_demand], color=['#1f77b4', '#ff7f0e', '#2ca02c'])
+    plt.title(f"Demand Validation in {country_code}, {horizon}")
+    plt.ylabel("Demand (TWh)", fontsize=12)
+    plt.xlabel("Data Source", fontsize=12)
+    plt.grid(axis='y', linestyle='--', alpha=0.7)
+    for bar in bars:
+        height = bar.get_height()
+        plt.text(bar.get_x() + bar.get_width() / 2, height, f'{height:.2f}',
+                 ha='center', va='bottom', fontsize=12)
+    plt.tight_layout()
+    plt.savefig(f"results/plots/demand_validation_ember_{country_code}.png")
+
+
+def plot_detailed_generation_validation(generation_df_, horizon, country_code):
+    ax = generation_df_.plot(kind="bar", figsize=(12, 7), width=0.8)
+    plt.title(
+        f"Comparison of Generation: EMBER vs PyPSA vs EIA, {horizon}", fontsize=16)
+    plt.xlabel("Generation Type", fontsize=12)
+    plt.ylabel("Generation (TWh)", fontsize=12)
+    ax.set_xticklabels(ax.get_xticklabels(), ha="right", fontsize=10)
+    plt.grid(True, which='both', axis='y', linestyle='--', linewidth=0.5)
+    plt.legend(["EMBER", "PyPSA", "EIA"], loc="upper right", fontsize=12)
+    plt.tight_layout()
+    plt.savefig(
+        f"results/plots/generation_validation_ember_{country_code}_detailed.png")
```

Comment:
My general thought is that this `data_validation.py` script can be placed into `plots` folder. We will keep `scripts` folder for pypsa-earth code related scripts (not visualization). It is just my thought, what do you think? The results of the validation then can be saved into `plots/results`.

Reply:
I think it is perfert. I will replace the folders in the right place.